apps/src/lib/tools/jsinterpreter/CustomMarshalingInterpreter.js

```diff
@@ -0,0 +1,272 @@
+   * @param {!Object} obj Data object.
+   * @param {*} name Name of property.
+   * @param {*} value New property value.
+   * @param {boolean} opt_fixed Unchangeable property if true.
+   * @param {boolean} opt_nonenum Non-enumerable property if true.
+   * @override
+   */
+  setProperty(
+    obj,
+    name,
+    value,
+    opt_fixed,
+    opt_nonenum
+  ) {
+    name = name.toString();
+    var nativeParent;
+    if (obj.isCustomMarshal) {
+      if (!this.shouldBlockCustomMarshalling_(name, obj)) {
+        obj.data[name] = codegen.marshalInterpreterToNative(this, value);
+      }
+    } else {
+      var hasProperty = false;
+      if (!obj.isPrimitive) {
+        hasProperty = super.hasProperty(obj, name);
+      }
+      if (!hasProperty &&
+          obj === this.globalScope &&
+          !!(nativeParent = this.customMarshaler.globalProperties[name]) &&
+          !this.shouldBlockCustomMarshalling_(name, obj, nativeParent)) {
+        nativeParent[name] = codegen.marshalInterpreterToNative(this, value);
+      } else {
+        return super.setProperty(obj, name, value, opt_fixed, opt_nonenum);
+      }
+    }
+  }
+
+  /**
+   * provides access to the original setProperty method to avoid
+   * custom marshaling in certain very specific cases.
+   */
+  setPropertyWithoutCustomMarshaling(...args) {
+    return super.setProperty(...args);
+  }
+
+
+  // The following overridden methods need to be patched in order to support custom marshaling.
+
+  // These changes revert a 10% speedup commit that bypassed hasProperty,
+  // getProperty, and setProperty:
+  // https://github.com/NeilFraser/JS-Interpreter/commit/c6f25b4a30046a858e5e90a92a8c0d24a93c0231
+
+  /**
+   * Retrieves a value from the scope chain.
+   * @param {!Object} name Name of variable.
+   * @return {!Object} The value.
+   * @override
+   */
+  getValueFromScope(name) {
+    var scope = this.getScope();
+    var nameStr = name.toString();
+    while (scope) {
+      if (this.hasProperty(scope, nameStr)) {
+        return this.getProperty(scope, nameStr);
+      }
+      scope = scope.parentScope;
+    }
+    this.throwException('Unknown identifier: ' + nameStr);
+    return this.UNDEFINED;
+  }
+
+  /**
+   * Sets a value to the current scope.
+   * @param {!Object} name Name of variable.
+   * @param {!Object} value Value.
+   * @param {boolean} declarator true if called from variable declarator.
+   * @override
+   */
+  setValueToScope(name, value, declarator) {
+    var scope = this.getScope();
+    var strict = scope.strict;
+    var nameStr = name.toString();
+    while (scope) {
+      if (this.hasProperty(scope, nameStr) || (!strict && !scope.parentScope)) {
+        if (declarator) {
+          // from a declarator, always call baseSetProperty
+          super.setProperty(scope, nameStr, value);
+        } else {
+          this.setProperty(scope, nameStr, value);
+        }
+        return;
+      }
+      scope = scope.parentScope;
```

Comment:
It really seems like we should have a test over this method walking up the scope chain.